domain/permission/state/state.go

```diff
@@ -0,0 +1,385 @@
+		} else if err != nil {
+			return errors.Annotatef(err, "adding permission %q for %q on %q", spec.Access, spec.User, spec.Target.Key)
+		}
+		return nil
+	})
+	if err != nil {
+		return corepermission.UserAccess{}, errors.Trace(domain.CoerceError(err))
+	}
+
+	userAccess.Access = spec.Access
+	userAccess.Object = objectTag(spec.Target)
+	return userAccess, nil
+}
+
+// findUser finds the user provided exists, hasn't been removed and is not
+// disabled. Return data needed to fill in corePermission.UserAccess.
+func findUser(
+	ctx context.Context,
+	tx *sqlair.TX,
+	userName string,
+) (User, error) {
+	var result User
+
+	getUserQuery := `
+SELECT (u.uuid, u.name, u.display_name, u.created_at, u.disabled) AS (&User.*),
+       creator.name AS &User.created_by_name
+FROM   v_user_auth u
+    LEFT JOIN user AS creator
+    ON        u.created_by_uuid = creator.uuid
+WHERE  u.removed = false AND u.name = $M.name`
+
+	selectUserStmt, err := sqlair.Prepare(getUserQuery, User{}, sqlair.M{})
+	if err != nil {
+		return result, errors.Annotate(err, "preparing select getUser query")
+	}
+
+	err = tx.Query(ctx, selectUserStmt, sqlair.M{"name": userName}).Get(&result)
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
+		return result, errors.Annotatef(usererrors.NotFound, "%q", userName)
+	} else if err != nil {
+		return result, errors.Annotatef(err, "getting user with name %q", userName)
+	}
+	if result.Disabled {
+		return result, errors.Annotatef(usererrors.AuthenticationDisabled, "%q", userName)
+	}
+	return result, nil
+}
+
+// validateTargetExists validates that the target of the permission
+// exists. An error is returned if not. Unless we have a controller
+// target, search for grant_on as a cloud.name and a model_list.uuid.
+// It must be one of those.
+func (s *State) validateTargetExists(
+	ctx context.Context,
+	tx *sqlair.TX,
+	targetKey string,
+) error {
+	if targetKey == coredatabase.ControllerNS {
+		return nil
+	}
+
+	// TODO: (hml) 6-Mar-24
+	// Add application offers check here when added to DDL.
+	targetExists := `
+SELECT &M.found_it FROM (
+    SELECT 1 AS found_it FROM cloud WHERE cloud.name = $M.grant_on
+    UNION
+    SELECT 1 AS found_it FROM model_list WHERE model_list.uuid = $M.grant_on
+)
+`
+	// Validate the grant_on target exists.
+	targetExistsStmt, err := sqlair.Prepare(targetExists, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Check that 1 row exists if the grant_on value is not ControllerNS.
+	// The behavior for this check is changing in sqlair, trying to
+	// account for either behavior. Old behavior: return no error and
+	// a slice length of zero. New behavior: if the length is 0, return
+	// ErrNoRows.
+	var foundIt = []sqlair.M{}
+	err = tx.Query(ctx, targetExistsStmt, sqlair.M{"grant_on": targetKey}).GetAll(&foundIt)
+	if err != nil {
+		return errors.Annotatef(err, "verifying %q target exists", targetKey)
+	}
+	if len(foundIt) != 1 {
```

Comment:
You might get as many rows as UNIONed statements. Check should be for non-zero.

Reply:
It's a bug if we get more than one row. These are unique values in the controller. There should never be 2 clouds with the same name, or 2 models with the same UUID.